omc3/harpy/clean.py

```diff
@@ -110,42 +116,81 @@ def _detect_bpms_with_spikes(bpm_data, max_peak_cut):
     if bpm_spikes.size:
-        LOGGER.debug(f"Spikes > {max_peak_cut} detected. BPMs removed: {bpm_spikes.size}")
+        LOGGER.debug(
+            f"Spikes > {max_peak_cut} detected. "
+            f"{bpm_spikes.size} BPMs removed: {', '.join(bpm_spikes)}"
+        )
     return bpm_spikes
 
 
+def _detect_bpms_with_nans(bpm_data) -> pd.Index:
+    """Detects BPMs with NaN values."""
+    nan_bpms = bpm_data[bpm_data.isna().any(axis=1)].index
+    if nan_bpms.size:
+        LOGGER.warning(
+            f"NaN BPMs detected. "
+            f"{nan_bpms.size} BPMs removed: {', '.join(nan_bpms)}"
+        )
+    return nan_bpms
+
+
 def _detect_bpms_with_exact_zeros(bpm_data, keep_exact_zeros):
     """Detects BPMs with exact zeros due to OP workaround."""
     if keep_exact_zeros:
         LOGGER.debug("Skipped exact zero check")
         return pd.Index([])
     exact_zeros = bpm_data[~np.all(bpm_data, axis=1)].index
     if exact_zeros.size:
-        LOGGER.debug(f"Exact zeros detected. BPMs removed: {exact_zeros.size}")
+        LOGGER.debug(
+            f"Exact zeros detected. "
+            f"{exact_zeros.size} BPMs removed: {', '.join(exact_zeros)} "
+        )
     return exact_zeros
 
 
-def _get_bad_bpms_summary(harpy_input, known_bad_bpms, bpm_flatness, bpm_spikes, exact_zeros):
-    return ([f"{bpm_name} Known bad BPM" for bpm_name in known_bad_bpms] +
-            [f"{bpm_name} Flat BPM, the difference between min/max is smaller than "
-            f"{harpy_input.peak_to_peak}" for bpm_name in bpm_flatness] +
-            [f"{bpm_name} Spiky BPM, found spike higher than "
-            f"{harpy_input.max_peak}" for bpm_name in bpm_spikes] +
-            [f"{bpm_name} Found an exact zero" for bpm_name in exact_zeros])
+def _get_bad_bpms_summary(harpy_input, **kwargs):
+    human_readable = {
+        "known": "Known bad BPM",
+        "flat": f"Flat BPM, the difference between min/max is smaller than {harpy_input.peak_to_peak}",
+        "spike": f"Spiky BPM, found spike higher than {harpy_input.max_peak}",
+        "zero": "Found an exact zero",
+        "nan": "Found NaN",
+    }
+
+    # Quick check that catches coding errors/typos:
+    unknown_kwargs = [kwarg for kwarg in kwargs if kwarg not in human_readable.keys()]
+    if len(unknown_kwargs):
+        raise NameError(f"Unknown reason(s) for Bad-BPMs: {unknown_kwargs}")
+
+    return [
+        f"{bpm_name} {msg}" for reason, msg in human_readable.items() for bpm_name in kwargs[reason]
+    ]
 
 
-def _report_clean_stats(n_total_bpms, n_good_bpms):
+def _report_clean_stats(n_total_bpms, n_good_bpms, bad_bpms_with_reasons):
     LOGGER.debug("Filtering done:")
+    
+    # As it not written out yet, provide more info in case of raising errors:
+    bad_bpms_message = "Bad BPMs found:\n"
+    bad_bpms_message += "\n".join(bad_bpms_with_reasons)
+    
+    # If all BPMs have been bad ---
     if n_good_bpms == 0:
+        LOGGER.info(bad_bpms_message)
```

Comment:
Nice

Reply:
yes, this was a pain to debug SPS measurements when it simply said that the bpms where bad without the why, even though it KNEW the why...